src/power_grid_model/data_types.py

```diff
@@ -7,84 +7,210 @@
 have been defined and explained in this file.
 """
 
-from typing import TypeVar, Union
+from typing import TypeAlias, TypeVar, Union
 
 import numpy as np
 
-from power_grid_model.core.dataset_definitions import ComponentTypeLike  # pylint: disable=unused-import
 from power_grid_model.core.dataset_definitions import ComponentTypeVar
 
+AttributeType: TypeAlias = str
+"""
+An attribute type is a string reprenting the attribute type of a specific component.
+
+- Examples:
+
+    - "id"
+    - "u_rated"
+"""
+
+SparseDataComponentType: TypeAlias = str
+"""
+A string representing the component type of sparse data structures.
+
+Must be either "data" or "indptr".
+"""
+
 SingleArray = Union[np.ndarray]
 """
-A single array is a one-dimensional structured containing a list of components of the same type.
+A single array is a one-dimensional structured numpy array containing a list of components of the same type.
 
 - Examples:
 
     - structure: <1d-array>
     - concrete: array([(0, 10500.0), (0, 10500.0)], dtype=power_grid_meta_data["input"]["node"].dtype)
 """
 
+SingleColumn = Union[np.ndarray]
+"""
+A single column is a one-dimensional structured numpy array containing a list of the same attribute of
+multiple components of the same type.
+
+- Examples:
+
+    - structure: <1d-array>
+    - concrete:
+
+        - array([0, 1], dtype=power_grid_meta_data["input"]["node"].dtype.fields["id"][0])
+        - array([10500.0, 10500.0], dtype=power_grid_meta_data["input"]["node"].dtype.fields["u_rated"][0])
+"""
+
+SingleColumnarData = dict[AttributeType, SingleColumn]
+"""
+Single columnar data is a dictionary where the keys are the attribute types of the same component
+and the values are :class:`SingleColumn`.
+
+- Example: {"id": :class:`SingleColumn`, "u_rated": :class:`SingleColumn`}
+"""
+
 DenseBatchArray = Union[np.ndarray]
 """
 A dense batch array is a two-dimensional structured numpy array containing a list of components of 
-the same type for each scenario.
+the same type for each scenario. Otherwise similar to :class:`SingleArray`.
```

Comment:
I feel confused here. How is the additional dimension handling the different scenarios? Because that is the only difference between `SingleArray`, right?

Reply:
Indeed, but we currently don't use full `numpy.typing` yet. That would indeed resolve a lot of the ambiguity. Unfortunately, it's quite tricky to use and will take some time to fully implement. Above all, not all python versions are supported (e.g.: we need Python 3.10 for it, which we only got recently).

the best we can do so far is make the distinction and let mypy and exceptions at runtime handle the rest

Reply:
Got it. So probably is a good idea to add a ticket far into the backlog (and/or github issue) to remind us to address this? Just adding `numpy.typing` and fixing little things like this one can improve readability a lot.

Reply:
it's not a "little thing" 😛 this is going to be some investigation to just try it out and then applying it. We can add a ticket to the backlog but I don't think it will be prioritized until we'll do some experiments with it, so there's not really a good reason to do it now.

Reply:
Didn't mean "little thing" in that way. But I understand.